backend/src/api.py

```diff
@@ -92,6 +95,128 @@ class NodeData:
 T = TypeVar("T", bound=RunFn)
 
 
+def validateTypes(
+    wrapped_func: RunFn,
+    schema_id: str,
+    py_var: str,
+    py_type: Union[str, Any],
+    associated_type: Type,
+):
+    evaluated_py_type = py_type
+    if isinstance(py_type, str):
+        try:
+            # Allows us to use pipe unions still
+            if "|" in py_type:
+                py_type = f'Union[{py_type.replace(" |", ",")}]'
+            # Gotta add these to the scope
+            local_scope = {
+                "Union": Union,
+                "np": np,
+            }
+            # pylint: disable=eval-used
+            evaluated_py_type = eval(
+                py_type,
+                wrapped_func.__globals__,
+                local_scope,
+            )
+        except Exception as e:
+            raise ValueError(
+                f"Unable to evaluate type for {schema_id}: {py_type=} | {e}"
+            ) from e
+    if evaluated_py_type is not associated_type:
+        if str(py_type).startswith("Union"):
+            evaluated_py_type_args = get_args(evaluated_py_type)
+            if associated_type not in evaluated_py_type_args:
+                raise ValueError(
+                    f"Type mismatch for {schema_id} (i='{py_var}'): {evaluated_py_type=} not in {evaluated_py_type_args=}"
+                )
+        elif str(associated_type).startswith("typing.Union"):
+            associated_type_args = get_args(associated_type)
+            if evaluated_py_type not in associated_type_args:
+                raise ValueError(
+                    f"Type mismatch for {schema_id} (i='{py_var}'): {evaluated_py_type=} not in {associated_type_args=}"
+                )
+        else:
+            raise ValueError(
+                f"Type mismatch for {schema_id} (i='{py_var}'): {evaluated_py_type=} != {associated_type=}"
+            )
+
+
+def typeValidateSchema(
+    wrapped_func: RunFn,
+    node_type: NodeType,
+    schema_id: str,
+    p_inputs: list[BaseInput],
+    p_outputs: list[BaseOutput],
+):
+    """Runtime validation for the number of inputs/outputs compared to the type args
+    While this isn't a comprehensive check, it's a good start for ensuring parity between the schema and the types
+    """
+
+    # We can't use inspect.get_annotations() since we need to support 3.8+
+    ann = getattr(wrapped_func, "__annotations__", None)
+
+    if ann is None:
+        return
+
+    # We don't want to run this check on iterator helpers as they can have different input/output metadata than what they actually run
+    if node_type == "iteratorHelper":
+        return
+
+    ### Return type validation
+
+    # Pop the return type key off the dict
+    return_type = ann.pop("return", None)
```

Comment:
I think this might be changing the dict associated with function. Please make a copy of the annotations so we don't change the function metadata.